JS: Add suffix/prefix construction to js/redos

erik-krogh opened the pull request:
And a bunch of performance improvements.   
The test case now runs faster than before I started the ReDoS work.   
(even though the query does way more work, and the testcase is bigger). 

I started with implementing [the `$` support as suggested by `asgerf`](https://github.com/github/codeql/pull/4585#discussion_r526162616).    
Then I had to somehow use that to remove FPs.   
That ended up being a search for a rejecting suffix.   
(I also tried a bit to get a concrete suffix, but that turned out to be hard).   

[An evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/eriks-profile-machine_1606115745487) shows some FPs that were removed, and no difference in performance.

asgerf commented on javascript/ql/src/Performance/ReDoS.ql:
```diff
@@ -861,44 +898,215 @@ predicate isPumpable(State fork, string w) {
+    hasRejectEdge(s, _)
     or
-    prev = process(fork, w, i - 1)
+    // all edges (at least one) with some char leads to another state that is rejectable.
```
Please update the predicate name and its documentation to clarify that this is a best-effort approximation, and does not generally provide the "definitely" guarantee the current name seems to suggest.

I cooked up a counter example, although after reading the test cases it seems you had already found similar FPs:
![graphviz (4)](https://user-images.githubusercontent.com/316427/100216810-a3d8b980-2f0a-11eb-960e-773ba7e9c2b9.png)


(All states here are accept states, or equivalently, have an epsilon-edge to the canonical accept state).

The states `A2` and `B2` are both rejectable (by `b` and `a`, respectively). They'll be recognized by the `hasRejectEdge` case.

The states `A1` and `B1` are rejectable as well, (by `ab` and `aa`, respectively). They'll be recognized by the recursive case, since reading `a` leads to (only) known definite reject states.

The state `S` is not rejectable, however, but will be recognized by the recursive case for the same reason as above: reading `a` leads only to definite rejectable states. But the two states we reach have disjoint sets of rejecting suffixes, so there is no suffix that can force rejection from `S`.

---

Finding a rejecting string in an NFA (or in a regexp) is NP-hard, so we can't realistically solve it exactly, and even a non-trivial "definitely" check is likely too hard.

erik-krogh replied:
I had some of the same thoughts.  

And it's actually even harder.   
After processing the `pump` string many times, the NFA will go through some number of states when processing the `pump` string one more time (the fixpoint of processing `pump`).  
For the suffix construction to be precise, we need to find a common suffix for all the states in this fixpoint  
(otherwise the backtracking will terminate quickly). 

> although after reading the test cases it seems you had already found similar FPs:

I haven't found concrete FPs related to this.   

> Finding a rejecting string in an NFA (or in a regexp) is NP-hard, so we can't realistically solve it exactly, and even a non-trivial "definitely" check is likely too hard.

Agreed.   

I could probably find concrete suffixes for some states, by doing a search that is conceptually similar to the prefix construction. But I think we would just get many FNs instead of a few FPs.

asgerf replied:
> I haven't found concrete FPs related to this.

Based on the comment below, I assumed this was caused by the suffix check.
```js
// GOOD - but we don't find that no suffix is rejected
var good32 = /(a+)*([^]{2,}|X)$/;
```
The double negation maybe have thrown me off. Actually I'm still struggling with that sentence 😕

erik-krogh replied:
Yeah, that sentence is bad.   

And what I meant is that i haven't found any FPs in the wild.  

The problem we have with that regexp is the range `[^]{2,}`, which accepts all strings of length 2 or more.   
We don't parse ranges when constructing the NFA. So the range is absent from the NFA. 

At the fixpoint of processing the `pump` string we will be in this catch-all range, and thus there is no rejecting suffix. 

With the NFA we do have (that doesn't include the range), a rejecting suffix is trivial to find.   
So we get a FP for that test.

erik-krogh replied:
(I've since found FPs related to the above in the wild).

asgerf commented on javascript/ql/src/Performance/ReDoS.ql:
```diff
@@ -861,44 +898,215 @@ predicate isPumpable(State fork, string w) {
+ * Holds if `term` may cause exponential backtracking on strings containing many repetitions of `witness`.
+ * Gets the minimum possible string that causes exponential backtracking.
+ */
+predicate isReDoSAttackable(RegExpTerm term, string witness, State s) {
```
I don't understand why this is called `witness`. Could we call it `pump` or something like that?

I'd say "witness" would be the three-tuple (prefix, pump, suffix).

asgerf reviewed (commented):
Nice!

The evaluation looks promising, though I think `/(?:\d(?:\s+|&nbsp;|&#160;)*){20}/i` was actually a TP:
```js
/(?:\d(?:\s+|&nbsp;|&#160;)*){20}/i.test('3' + ' '.repeat(100)) // hangs
```
Did we lose that because we don't handle fixed-repetition quantifiers the same way? (the string is rejectable because there has to be 20 digits in the string, but you can pump before getting to the second digit).

erik-krogh commented:
> The evaluation looks promising, though I think `/(?:\d(?:\s+|&nbsp;|&#160;)*){20}/i` was actually a TP:
> 
> ```js
> /(?:\d(?:\s+|&nbsp;|&#160;)*){20}/i.test('3' + ' '.repeat(100)) // hangs
> ```

Nice catch.  

That was caused by [this optimization](https://github.com/github/codeql/pull/4721/commits/88fc2d361a73291c37c0d840a2695fbbc8609583), which is apparently unsound, and I didn't catch that. 

I've rebased that commit away.

asgerf commented:
Could you verify that performance is still reasonable after reverting that optimization? I'd be happy with just timing the test suite (which is expensive for this query) and the projects that had results in the last run.

erik-krogh commented:
> Could you verify that performance is still reasonable after reverting that optimization? I'd be happy with just timing the test suite (which is expensive for this query) and the projects that had results in the last run.

The test-suite is fine. 

[A evaluation (on the projects with results from the last run)](https://git.semmle.com/erik/dist-compare-reports/tree/eriks-profile-machine_1606400292105) definitely shows a performance regression.   
But not a bad one.

erik-krogh commented:
> [A evaluation (on the projects with results from the last run)](https://git.semmle.com/erik/dist-compare-reports/tree/eriks-profile-machine_1606400292105) definitely shows a performance regression.
> But not a bad one.

I got [some improvements in the pipeline that fixes the performance regression](https://github.com/github/codeql/compare/nextReDoS...erik-krogh:starReDoS?expand=1), and also finds even more results.   

I got a large evaluation running, but [preliminary results looks promising](https://git.semmle.com/erik/dist-compare-reports/tree/erik-laptop_1606478389468)   
(all new results are TPs, except `/(a?)*b/`).

erik-krogh commented:
A [larger evaluation looked good](https://git.semmle.com/erik/dist-compare-reports/tree/eriks-profile-machine_1606491070237) (the baseline is the [`update expected output`](https://github.com/github/codeql/pull/4721/commits/9468a6e8dc7771e96ca2957a5db899bc77c1f65f) commit).   
So I've added the improvements I had in the pipeline to this PR.   
There is lots of new TPs, and 2 FPs (the one mentioned above, and another one related to suffix checking). 

I found [this optimization](https://github.com/github/codeql/pull/4721/commits/8a3e87fe42b0e642dec2fdd64ca7aaea4bb1ae49) while debugging another thing.  
Turns out we don't need that part I removed in that commit. It makes no difference in results across `full.slugs`.

erik-krogh commented:
I added [a refinement](https://github.com/github/codeql/pull/4721/commits/33b2701551633c784b5fbcfd8546f719096e4011) to get the FP count down to 0.   

The `concretise` predicate does not produce the correct pump string in all cases.    
But I'm somewhat sure I've removed the possible false positives.

erik-krogh commented:
[An evaluation on just the last commit](https://git.semmle.com/erik/dist-compare-reports/tree/eriks-profile-machine_1606750570306) confirms that the only change in results is removing the single FP I had found.

asgerf commented on javascript/ql/src/Performance/ReDoS.ql:
```diff
@@ -705,7 +705,25 @@ predicate isFork(State q, InputSymbol s1, InputSymbol s2, State r1, State r2) {
     q1 = q2 and
-    epsilonSucc+(q) = q
+    epsilonSucc+(q) = q and
+    exists(RegExpTerm term | term = q.getRepr() | term instanceof InfiniteRepetitionQuantifier) and
```
I'm not seeing the rationale behind this whole epsilon-loop case.

RegExp engines will generally not allow the body of a repetition clause to match the empty string. Attempting to match `E*` while allowing `E` to match the empty string immediately leads to an infinite loop.

A fork that matches a loop of epsilon edges therefore looks suspicious to me, as this loop is likely a spurious artifact of our own NFA construction.

I've been trying to think of what cases we could be missing here that would explain the FNs you saw. Here's an idea:

Could it be that there are two distinct epsilon-paths (not necessarily looping) leading from `q` to `q1=q2`, and if the engine failed after trying the first path, it will retry using the other path? (which is futile of course, as it's in the same state, but the engine doesn't check for this).

(When I say _distinct_ paths, it's possible they actually need to be _disjoint_ paths.)

erik-krogh replied:
It's to flag something like `/(a?b?)*c/` ([debug session](https://regex101.com/r/30l4PM/2/debugger)).   
Where the body inside the repetition does match the empty-string.    
This causes the regexp engine to try to match against every possible subset of the input string (see the debug session).  

I'll look more into it, but I don't see another way to flag it.

erik-krogh replied:
I think the general pattern is that the body of the repetition can match something of different lengths, where there is no way for the engine to determine where one repetition starts/ends compared to the next repetition.    

In the above example, the body matches something of length between 0 and 2. 

Consider this example (which we don't flag): `var bad = /(a(\w)?)*X/`.   
The inner body either matches `a`, or `aa`.  
But these combinations of strings of length 1 or 2 can be combined in `O(2^n)` possible ways, giving us exponential backtracking.     

At least one of the regexp [just fixed in highlight.js](https://github.com/highlightjs/highlight.js/commit/373b9d862401162e832ce77305e49b859e110f9c#diff-5819541c636e70dcfc768352222f083986ce06ab582b657dca8cb50364e2a334L34) was of this pattern.

erik-krogh replied:
I'll revert the epsilon idea for this PR.   

I'll look into it for (yet another) followup PR.

asgerf replied:
Ok, I think I see how it works for `/(a?b?)*c/`. The fork state is right after reading `a`, as there are two different ways to read `b`: one by skipping `b?` and looping around, and the other by immediately consuming `b?`, but we missed that fork because both lead to the same NFA state. I can see the relevance of epsilon loop now. I'm still not sure if it has to be a loop or just two distinct paths, but matching loops as a safe approximation is fine with me.

I'm still not sure about the second commit that added the extra conditions.

I'll leave it up to you if you'd prefer to use this PR or in another one.

asgerf replied:
Actually maybe we're on to something here. The FPs you observed from the _first_ version of this check could be from the following spurious fork: Before reading `a?` there are seemingly two ways to read `a`: skipping `a?`and looping around; or immediately consuming `a?`. Exactly as with the `b` fork, except this one is spurious because it involves a repetition body matching an empty string, which won't ever happen.

erik-krogh replied:
> I'll leave it up to you if you'd prefer to use this PR or in another one.

I'll wait until another PR.   
I also found another thing I want to put in the next PR. 

> Ok, I think I see how it works for /(a?b?)*c/.   

I'm still not sure.   
When I try it out, the attack string must be repetitions of `ab`, otherwise the attack doesn't work.   
Just repeating `a` or `b` doesn't work.   

> I'm still not sure about the second commit that added the extra conditions.

Me too. But it was the best I could come up with.

erik-krogh commented:
@asgerf any more comments?  

Otherwise I would like to merge this before I open the next ReDoS PR.

asgerf reviewed (approved):
LGTM
